doschema/jsvalidation.py

```diff
@@ -0,0 +1,232 @@
+                            path = props[prop].get('$ref')
+                            props[prop] = self.resolver.resolve(path)[1]
+
+                    prop_type = props[prop].get('type')
+
+                    self.process_type_field(
+                        prop_type,
+                        curr_field + (prop, )
+                    )
+
+                    if prop_type == 'object':
+                        self.traverse(
+                            props[prop],
+                            curr_field + (prop, )
+                        )
+                    elif prop_type == 'array':
+                        self.process_array(
+                            props[prop]['items'],
+                            curr_field
+                        )
+
+            if check_dependencies and 'dependencies' in curr_schema:
+                self.traverse_dependencies(curr_schema, curr_field)
+
+        else:
+            raise ValueError('Curr_schema type not an object')
+
+    def process_array(self, field_value, curr_field):
+        """Process each JSON object in the array according to ignore_index.
+
+        :param field_value: Type of the field.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        field_path = curr_field
+        for elem in field_value:
+            if self.ignore_index:
+                new_field_path = field_path
+            else:
+                new_field_path = field_path + (field_value.index(elem), )
+
+            self.traverse(elem, new_field_path)
+
+    def process_type_field(self, field_value, curr_field):
+        """Check if a name field with type "type" has already appeared.
+
+        If had appeared with different type,raise an exception.
+        Otherwise, add object to a dict of types.
+
+        :param field_value: Type of the field.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        if curr_field in self.fields_types_dict.keys():
+            if self.fields_types_dict[curr_field].field_type != field_value:
+                err_msg = "Type mismatch in schemas {0} and {1}."
+                old_schema = self.fields_types_dict[curr_field].schema_index
+                raise JSONSchemaCompatibilityException(
+                    old_schema,
+                    self.uri,
+                    err_msg.format(old_schema, self.uri)
+                )
+        else:
+            self.fields_types_dict[curr_field] = FieldToAdd(
+                schema_index=self.uri,
+                field_tuple=curr_field,
+                field_type=field_value
+            )
+
+    def check_special(self, curr_schema, curr_field):
+        """Check schema for allOf, anyOf and oneOf.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        keys = self.COLLECTION_KEYS.intersection(curr_schema)
+        for key in keys:
```

Comment:
is it possible to have multiple collection key at the same time? If not we should raise an exception when `len(keys) > 1`